arcjet/index.ts

```diff
@@ -836,131 +792,114 @@ export function validateEmail(
-      return;
-    }
-  }
-}
-
 export function detectBot(
   options?: BotOptions,
   ...additionalOptions: BotOptions[]
 ): Primitive {
   const rules: ArcjetBotRule<{}>[] = [];
 
   // Always create at least one BOT rule
-  for (const opt of [options ?? {}, ...additionalOptions]) {
+  for (const opt of [options ?? { allow: [] }, ...additionalOptions]) {
     const mode = opt.mode === "LIVE" ? "LIVE" : "DRY_RUN";
-    // TODO: Filter invalid bot types (or error??)
-    const block = Array.isArray(opt.block)
-      ? opt.block
-      : [ArcjetBotType.AUTOMATED];
-    // TODO: Does this avoid prototype pollution by putting in a Map first?
-    const addMap = new Map();
-    for (const [key, value] of Object.entries(opt.patterns?.add ?? {})) {
-      addMap.set(key, value);
+    if (typeof opt.allow !== "undefined" && typeof opt.deny !== "undefined") {
+      throw new Error("Both allow and deny cannot be provided to detectBot");
+    }
+
+    let config: BotConfig = {
+      tag: "allowed-bot-config",
+      val: {
+        entities: [],
+        skipCustomDetect: true,
+      },
+    };
+    if (Array.isArray(opt.allow)) {
+      for (const allow of opt.allow) {
+        if (typeof allow !== "string") {
+          throw new Error("all values in `allow` must be a string");
+        }
+      }
+
+      config = {
+        tag: "allowed-bot-config",
+        val: {
+          entities: opt.allow,
+          skipCustomDetect: true,
+        },
+      };
+    }
+
+    if (Array.isArray(opt.deny)) {
+      for (const deny of opt.deny) {
+        if (typeof deny !== "string") {
+          throw new Error("all values in `allow` must be a string");
+        }
+      }
+
+      config = {
+        tag: "denied-bot-config",
+        val: {
+          entities: opt.deny,
+          skipCustomDetect: true,
+        },
+      };
     }
-    // TODO(#217): Additional validation on these `patterns` options
-    const add = Array.from(addMap.entries());
-    const remove = opt.patterns?.remove ?? [];
 
     rules.push({
       type: "BOT",
       priority: Priority.BotDetection,
       mode,
-      block,
-      add,
-      remove,
+      allow: Array.isArray(opt.allow) ? opt.allow : [],
+      deny: Array.isArray(opt.deny) ? opt.deny : [],
 
       validate(
         context: ArcjetContext,
         details: Partial<ArcjetRequestDetails>,
       ): asserts details is ArcjetRequestDetails {
-        assert(
-          typeof details.headers !== "undefined",
-          "DetectBot requires `headers` to be set.",
-        );
+        if (typeof details.headers === "undefined") {
+          throw new Error("bot detection requires `headers` to be set");
+        }
+        if (typeof details.headers.has !== "function") {
+          throw new Error(
+            "bot detection requires `headers` to extend `Headers`",
+          );
+        }
+        if (!details.headers.has("user-agent")) {
+          throw new Error("bot detection requires user-agent header");
```

Comment:
How would this be presented to the user?
I feel like they'd want to handle this differently from the two cases above since would both most likely be triggered by misuse of the SDK, but its completely valid that they'd receive a request with no user-agent header and might want to then decide whether to allow it or not.

Reply:
This gets turned into an ArcjetErrorReason. If we don't have a user agent, we can't detect bots so it's an error reason and we fail open.

Reply:
Yeah but how would a user use that? I would imagine that an error like the above two would be turned in to a 500, where not sending a user-agent header would suit a 400 better. If it became a 5xx that would be wrong since the error was caused by the client not sending required headers, not a server error.

Reply:
Yeah I think that structured errors would solve it quite nicely, but they are obviously out of scope for this. It might just be worth documenting it when it comes to updating the docs.

Reply:
https://github.com/arcjet/arcjet-docs/issues/45